url1a745d67fc_sobercheg_InterviewTraining_tgz-pJ8-leetcode_oj_Solution2TestJ8/src/leetcode/oj/Solution.java

```diff
@@ -0,0 +1,3742 @@
+        List<List<Integer>> nChooseKIndices = choose(num.length, k / 2);
+        Map<Integer, List<List<Integer>>> sumToCombinations = new HashMap<Integer, List<List<Integer>>>();
+        for (List<Integer> indices : nChooseKIndices) {
+            int sum = 0;
+            for (int index : indices) {
+                sum += num[index];
+            }
+            if (!sumToCombinations.containsKey(sum)) {
+                sumToCombinations.put(sum, new ArrayList<List<Integer>>());
+            }
+            sumToCombinations.get(sum).add(indices);
+        }
+
+        Set<List<Integer>> kSums = new HashSet<List<Integer>>();
+
+        // Step 2. Check whether S contains both some number x and its sibling target−x
+        if (k % 2 == 0) {
+            for (int sum : sumToCombinations.keySet()) {
+                int antiSum = target - sum;
+                if (sumToCombinations.containsKey(antiSum)) {
+                    List<List<Integer>> finalCombos = new ArrayList<List<Integer>>();
+                    for (int i = 0; i < sumToCombinations.get(sum).size(); i++) {
+                        for (int j = sum == antiSum ? i + 1 : 0; j < sumToCombinations.get(antiSum).size(); j++) {
+                            List<Integer> finalCombo = new ArrayList<Integer>();
+                            finalCombo.addAll(sumToCombinations.get(sum).get(i));
+                            finalCombo.addAll(sumToCombinations.get(antiSum).get(j));
+                            if (new HashSet<Integer>(finalCombo).size() < finalCombo.size()) continue;
+                            finalCombos.add(finalCombo);
+                        }
+                    }
+
+                    for (List<Integer> indices : finalCombos) {
+                        List<Integer> finalValues = new ArrayList<Integer>();
+                        for (int index : indices) {
+                            finalValues.add(num[index]);
+                        }
+                        Collections.sort(finalValues);
+                        kSums.add(finalValues);
+                    }
+//                    break;
+                }
+            }
+
+        } else {
+            //  For each input element z, check whether S contains both x and target+z−x, for some number x.
+            for (int z = 0; z < num.length; z++) {
+                for (int sum : sumToCombinations.keySet()) {
+                    int antiSum = target - sum - num[z];
+                    if (sumToCombinations.containsKey(antiSum)) {
+                        List<List<Integer>> finalCombos = new ArrayList<List<Integer>>();
+                        for (int i = 0; i < sumToCombinations.get(sum).size(); i++) {
+                            for (int j = (sum == antiSum) ? i + 1 : 0; j < sumToCombinations.get(antiSum).size(); j++) {
+                                List<Integer> finalCombo = new ArrayList<Integer>();
+                                finalCombo.addAll(sumToCombinations.get(sum).get(i));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.